shared/js/background/newtab-tracker-stats.js

```diff
@@ -0,0 +1,348 @@
+     * When the evictExpired alarm fires:
+     *   - cleanup data structure
+     *   - store cleaned data
+     *   - publish the changed data
+     * @param {number} [now] - optional timestamp to use in comparisons
+     */
+    handlePruneAlarm (now = Date.now()) {
+        this.stats.evictExpired()
+        this.syncToStorage()
+        this.sendToNewTab('following a evictExpired alarm')
+    }
+
+    /**
+     * Convert the internal data into a format that can be used in the New Tab Page UI
+     *
+     * First, we group & sort the data, and then we increase the count of "Other" to account
+     * for overflows or for trackers where the owner was not in the top 100 list
+     *
+     * @param {number} maxCount
+     * @param {number} [now] - optional timestamp to use in comparisons
+     * @returns {import("zod").infer<typeof import("../newtab/schema").dataFormatSchema>}
+     */
+    toDisplayData (maxCount = 10, now = Date.now()) {
+        // access the entries once they are sorted and grouped
+        const stats = this.stats.sorted(maxCount, now)
+
+        // is there an entry for 'unknownCompanyKey' (meaning an entry where the company name was skipped)
+        const index = stats.results.findIndex(result => result.key === NewTabTrackerStats.unknownCompanyKey)
+
+        // if there is an entry, add the 'overflow' count and move it to the end of the list
+        if (index > -1) {
+            const element = stats.results[index]
+            if (stats.overflow) {
+                element.count += stats.overflow
+            }
+            const spliced = stats.results.splice(index, 1)
+            stats.results.push(...spliced)
+        } else {
+            // if we get here, there was no entry for `unknownCompanyKey`, so we need to add one to cover the overflow
+            if (stats.overflow) {
+                stats.results.push({
+                    key: NewTabTrackerStats.unknownCompanyKey,
+                    count: stats.overflow
+                })
+            }
+        }
+
+        // now produce the data in the shape consumers require for rendering their UI
+        return {
+            totalCount: this.stats.totalCount,
+            totalPeriod: 'install-time',
+            trackerCompaniesPeriod: 'last-hour',
+            trackerCompanies: stats.results.map(item => {
+                // convert our known key into the 'Other'
+                const displayName = item.key === NewTabTrackerStats.unknownCompanyKey
+                    ? 'Other'
+                    : item.key
+
+                return {
+                    displayName,
+                    count: item.count
+                }
+            })
+        }
+    }
+
+    debounceTimers = {}
+
+    /**
+     * @param {string} name
+     * @param {number} timeout
+     * @param {() => unknown} fn
+     */
+    debounced (name, timeout, fn) {
+        clearTimeout(this.debounceTimers[name])
+        this.debounceTimers[name] = setTimeout(fn, timeout)
```

Comment:
This debounce implementation seems to delay indefinitely when new events come in regularly. e.g. if there is a tracker that is making a request on a 500ms interval, we will keep debouncing that update, blocking the update of the new tab page until that tab is closed.

Reply:
thanks for catching this, @sammacbeth!

I've moved to a trailing throttle implementation instead, so that events will always occur at most once during a time window.